### PR TITLE
Start/Stop API initial commit

### PR DESCRIPTION
**Pull request by lobdelle**
*Issue #, if available:*
Start/Stop API

*Description of changes:*
Added Start and Stop API classes and tests. 

Currently have some failing tests (commented out) that depend on the expected behavior of a running transform job.

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformRequest.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform.action.start
+
+import org.elasticsearch.action.ActionRequestValidationException
+import org.elasticsearch.action.ValidateActions.addValidationError
+import org.elasticsearch.action.update.UpdateRequest
+import org.elasticsearch.common.io.stream.StreamInput
+import org.elasticsearch.common.io.stream.StreamOutput
+import java.io.IOException
+
+class StartTransformRequest : UpdateRequest {
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformRequest.kt:10]: nit: I think you can do inline constructor 

```
class StartTransformRequest(val id: String) : UpdateRequest(id) {
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    transportService: TransportService,
+    val client: Client,
+    actionFilters: ActionFilters
+) : HandledTransportAction<StartTransformRequest, AcknowledgedResponse>(
+    StartTransformAction.NAME, transportService, actionFilters, ::StartTransformRequest
+) {
+    private val log = LogManager.getLogger(javaClass)
+
+    override fun doExecute(task: Task, request: StartTransformRequest, actionListener: ActionListener<AcknowledgedResponse>) {
+        val getReq = GetTransformRequest(request.id(), null)
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:43]: minor: I see default values are set for srcContext in ``GetTransformRequest`` you don't need to explicitly pass ``null``. You can just do `GetTransformRequest(request.id())`

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform.action.stop
+
+import org.elasticsearch.action.ActionRequestValidationException
+import org.elasticsearch.action.ValidateActions.addValidationError
+import org.elasticsearch.action.update.UpdateRequest
+import org.elasticsearch.common.io.stream.StreamInput
+import org.elasticsearch.common.io.stream.StreamOutput
+import java.io.IOException
+
+class StopTransformRequest : UpdateRequest {
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt:10]: nit: same as start can have inline constructor

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform.action.stop
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt:1]: missing license

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformAction.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform.action.start
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformAction.kt:1]: looks all new files are missing lines, can you add at all places?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    private fun getTransformMetadata(transform: Transform, actionListener: ActionListener<AcknowledgedResponse>) {
+        val req = GetRequest(INDEX_MANAGEMENT_INDEX, transform.metadataId).routing(transform.id)
+        client.get(req, object : ActionListener<GetResponse> {
+            override fun onResponse(response: GetResponse) {
+                if (!response.isExists || response.isSourceEmpty) {
+                    // If there is no metadata doc then the runner will instantiate a new one
+                    // in FAILED status which the user will need to retry from
+                    actionListener.onResponse(AcknowledgedResponse(true))
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:107]: Just thinking through this, during transform run we always first create transformMetadata and then associate to the transform document. So if an metadata id exists in transform but there is document then the only reason will be someone delete the transform metadata manually. I think we should just throw an exception here metadata is in weird state and cannot be started

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                } else {
+                    val metadata = response.sourceAsBytesRef?.let {
+                        val xcp = XContentHelper.createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE, it, XContentType.JSON)
+                        xcp.parseWithType(response.id, response.seqNo, response.primaryTerm, TransformMetadata.Companion::parse)
+                    }
+                    if (metadata == null) {
+                        // If there is no metadata doc then the runner will instantiate a new one
+                        // in FAILED status which the user will need to retry from
+                        actionListener.onResponse(AcknowledgedResponse(true))
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:116]: Same as above this should be an exception

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                }
+            }
+
+            override fun onFailure(e: Exception) {
+                actionListener.onFailure(ExceptionsHelper.unwrapCause(e) as Exception)
+            }
+        })
+    }
+
+    private fun getTransformMetadata(transform: Transform, actionListener: ActionListener<AcknowledgedResponse>) {
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:100]: This is a little confusing name, the method is actually attempting to fetch and updateTransformMetadata not just fetching transformMetadata.

Do you think it makes sense to rename this to something along lines of `retrieveAndUpdateTransformMetadata`?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        })
+    }
+
+    private fun updateTransformMetadata(transform: Transform, metadata: TransformMetadata, actionListener: ActionListener<AcknowledgedResponse>) {
+        val now = Instant.now().toEpochMilli()
+        val updatedStatus = when (metadata.status) {
+            TransformMetadata.Status.FINISHED, TransformMetadata.Status.STOPPED -> TransformMetadata.Status.STARTED
+            TransformMetadata.Status.STARTED, TransformMetadata.Status.INIT ->
+                return actionListener.onResponse(AcknowledgedResponse(true))
+            TransformMetadata.Status.FAILED -> TransformMetadata.Status.FAILED // What should happen here?
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:135]: The status will be set to `STARTED` and we should also clear the metadata `afterKey` field

Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:135]: I think for now we can just move to `STARTED`. When we have incremental transforms we can see if a new status is required

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                }
+            }
+
+            override fun onFailure(e: Exception) {
+                actionListener.onFailure(ExceptionsHelper.unwrapCause(e) as Exception)
+            }
+        })
+    }
+
+    private fun getTransformMetadata(transform: Transform, request: StopTransformRequest, actionListener: ActionListener<AcknowledgedResponse>) {
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt:81]: do you think renaming this makes sense?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                actionListener.onFailure(ExceptionsHelper.unwrapCause(e) as Exception)
+            }
+        })
+    }
+
+    private fun updateTransformJob(transform: Transform, request: StopTransformRequest, actionListener: ActionListener<AcknowledgedResponse>) {
+        val now = Instant.now().toEpochMilli()
+        request.index(IndexManagementPlugin.INDEX_MANAGEMENT_INDEX).setIfSeqNo(transform.seqNo).setIfPrimaryTerm(transform.primaryTerm)
+            .doc(mapOf(Transform.TRANSFORM_TYPE to mapOf(Transform.ENABLED_FIELD to false,
+                Transform.ENABLED_AT_FIELD to null, Transform.ENABLED_AT_FIELD to now)))
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt:149]: duplication of Transform.ENABLED_AT_FIELD

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformRequest.kt (hunk 1 above)
Comment (lobdelle) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/StartTransformRequest.kt:10]: The above doesn't work, as it's confusing the constructor being referenced. It's trying to use the StreamInput constructor and throwing a type mismatch error on `UpdateRequest(id)`, as UpdateRequest doesn't have an ID only constructor. Currently I think it's using the no-arg constructor, then updating the ID and Index after initialization.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt (hunk 3 above)
Comment (lobdelle) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/StopTransformRequest.kt:10]: See notes for start

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt (hunk 11 above)
Comment (lobdelle) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt:149]: proper field corrected to UPDATED_AT_FIELD

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        })
+    }
+
+    private fun retrieveAndUpdateTransformMetadata(transform: Transform, actionListener: ActionListener<AcknowledgedResponse>) {
+        val req = GetRequest(INDEX_MANAGEMENT_INDEX, transform.metadataId).routing(transform.id)
+        client.get(req, object : ActionListener<GetResponse> {
+            override fun onResponse(response: GetResponse) {
+                if (!response.isExists || response.isSourceEmpty) {
+                    // If there is no metadata doc then the runner will instantiate a new one
+                    // in FAILED status which the user will need to retry from
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:121]: nit: can you remove this?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    ) {
+        val now = Instant.now().toEpochMilli()
+        val updatedStatus = when (metadata.status) {
+            TransformMetadata.Status.STARTED, TransformMetadata.Status.INIT, TransformMetadata.Status.STOPPED -> TransformMetadata.Status.STOPPED
+            TransformMetadata.Status.FINISHED, TransformMetadata.Status.FAILED -> metadata.status
+        }
+
+        val updateRequest = UpdateRequest(IndexManagementPlugin.INDEX_MANAGEMENT_INDEX, transform.metadataId)
+            .doc(mapOf(TransformMetadata.TRANSFORM_METADATA_TYPE to mapOf(TransformMetadata.STATUS_FIELD to updatedStatus.type,
+                TransformMetadata.LAST_UPDATED_AT_FIELD to now)))
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/stop/TransportStopTransformAction.kt:144]: missing `.routingId(transform.id)`

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        val now = Instant.now().toEpochMilli()
+        val updatedStatus = when (metadata.status) {
+            TransformMetadata.Status.FINISHED, TransformMetadata.Status.STOPPED -> TransformMetadata.Status.STARTED
+            TransformMetadata.Status.STARTED, TransformMetadata.Status.INIT ->
+                return actionListener.onResponse(AcknowledgedResponse(true))
+            TransformMetadata.Status.FAILED -> TransformMetadata.Status.STARTED
+        }
+        val updateRequest = UpdateRequest(INDEX_MANAGEMENT_INDEX, transform.metadataId)
+            .doc(mapOf(TransformMetadata.TRANSFORM_METADATA_TYPE to mapOf(TransformMetadata.STATUS_FIELD to updatedStatus.type,
+                TransformMetadata.FAILURE_REASON to null, TransformMetadata.LAST_UPDATED_AT_FIELD to now)))
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/start/TransportStartTransformAction.kt:156]: missing `.routingId(transform.id)`

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStartTransformActionIT.kt [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    fun `test start a transform with no id fails`() {
+        try {
+            client().makeRequest("POST", "$TRANSFORM_BASE_URI//_start")
+            fail("Expected 400 Method BAD_REQUEST response")
+        } catch (e: ResponseException) {
+            assertEquals("Unexpected status", RestStatus.BAD_REQUEST, e.response.restStatus())
+        }
+    }
+
+    /*
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStartTransformActionIT.kt:80]: can you uncomment these? 
Also can you add testcase for attempting to start a transform that doesn't exist?

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+        // Assert transform still failed status
+        waitFor {
+            val updatedTransform = getTransform(transform.id)
+            val metadata = getTransformMetadata(updatedTransform.metadataId!!)
+            assertEquals("Transform should have stayed failed", TransformMetadata.Status.FAILED, metadata.status)
+        }
+    }
+
+    /* Goes straight to finished before test can pick up STARTED status
+    @Throws(Exception::class)
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt:144]: uncomment? And add a test where transform is missing

**Comment by lobdelle** [comment]
Stop Transform API Tests failing intermittently on build. Tests pass consistently when run independently, but failing on build.

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+            )
+        ).let { createTransform(it, it.id) }
+
+        updateTransformStartTime(transform)
+
+        waitFor {
+            val transformJob = getTransform(transformId = transform.id)
+            assertNotNull("Transform job doesn't have metadata set", transformJob.metadataId)
+            val transformMetadata = getTransformMetadata(transformJob.metadataId!!)
+            assertEquals("Transform is not STARTED", TransformMetadata.Status.STARTED, transformMetadata.status)
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt:173]: I am seeing these errors in the workflow ``java.lang.AssertionError: Transform is not STARTED expected:<started> but was:<finished>``, probably job is completed by the time waitFor block is able to find the metadata associated with the job. 

Can you try to change the transform job so it runs longer?
```
generateNYCTaxiData("source_test_stopping_running_transform")
val transform = randomTransform().copy(
            id = "test_stopping_running_transform",
            schemaVersion = 1L,
            enabled = true,
            jobSchedule = IntervalSchedule(Instant.now(), 1, ChronoUnit.MINUTES),
            updatedAt = Instant.now(),
            enabledAt = Instant.now(),
            description = "basic search test",
            sourceIndex = "source"_test_stopping_running_transform",
            targetIndex = "target_test_stopping_running_transform",
            metadataId = null,
            roles = emptyList(),
            pageSize = 1,
            groups = listOf(
                Terms(sourceField = "tpep_pickup_datetime", targetField = "pickup time", period="60m"),
                Terms(sourceField = "store_and_fwd_flag", targetField = "flag")
            )
        ).let { createTransform(it, it.id) }
```

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStartTransformActionIT.kt [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+import java.time.temporal.ChronoUnit
+
+@TestLogging(value = "level:DEBUG", reason = "Debugging tests")
+@Suppress("UNCHECKED_CAST")
+class RestStartTransformActionIT : TransformRestTestCase() {
+
+    @Throws(Exception::class)
+    fun `test starting a stopped transform`() {
+        val transform = createTransform(randomTransform().copy(enabled = false, enabledAt = null, metadataId = null))
+        assertTrue("Transform was not disabled", !transform.enabled)
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStartTransformActionIT.kt:39]: nit: could do ``assertFalse("Transform was not disabled", transform.enabled)``

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+            val updatedTransform = getTransform(transform.id)
+            val metadata = getTransformMetadata(updatedTransform.metadataId!!)
+            assertEquals("Transform never failed", TransformMetadata.Status.FAILED, metadata.status)
+        }
+
+        // Stop transform
+        val response = client().makeRequest("POST", "$TRANSFORM_BASE_URI/${transform.id}/_stop")
+        assertEquals("Stop transform failed", RestStatus.OK, response.restStatus())
+        val expectedResponse = mapOf("acknowledged" to true)
+        assertEquals(expectedResponse, response.asMap())
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt:134]: Can you move this inside waitFor{} block, I am seeing transient failures for test too when I run the test in a loop

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+import org.elasticsearch.test.junit.annotations.TestLogging
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+@TestLogging(value = "level:DEBUG", reason = "Debugging tests")
+@Suppress("UNCHECKED_CAST")
+class RestStopTransformActionIT : TransformRestTestCase() {
+
+    @Throws(Exception::class)
+    fun `test stopping a started transform`() {
```
[the remark this answers is not in the document]
Comment (lobdelle) [reply src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestStopTransformActionIT.kt:38]: Removed

**Review by thalurur — APPROVED** [review]

**Pull request merged**